src/group/expand.rs

```diff
@@ -7,71 +7,83 @@
+use generic_array::{
+    sequence::Concat,
+    typenum::{Unsigned, U1, U2},
+    ArrayLength, GenericArray,
+};
 
 // Computes ceil(x / y)
 fn div_ceil(x: usize, y: usize) -> usize {
     let additive = (x % y != 0) as usize;
     x / y + additive
 }
 
-fn xor(x: &[u8], y: &[u8]) -> Result<Vec<u8>, InternalError> {
-    if x.len() != y.len() {
-        return Err(InternalError::HashToCurveError);
-    }
-
-    Ok(x.iter().zip(y).map(|(&x1, &x2)| x1 ^ x2).collect())
+fn xor<L: ArrayLength<u8>>(x: GenericArray<u8, L>, y: GenericArray<u8, L>) -> GenericArray<u8, L> {
+    x.into_iter().zip(y).map(|(x1, x2)| x1 ^ x2).collect()
 }
 
 /// Corresponds to the expand_message_xmd() function defined in
 /// <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.txt>
-pub fn expand_message_xmd<H: BlockInput + Digest>(
+pub fn expand_message_xmd<
+    H: BlockInput + Digest,
+    L: ArrayLength<u8>,
+    D: ArrayLength<u8> + Add<U1>,
+>(
     msg: &[u8],
-    dst: &[u8],
-    len_in_bytes: usize,
-) -> Result<Vec<u8>, InternalError> {
-    let ell = div_ceil(len_in_bytes, <H as Digest>::OutputSize::USIZE);
+    dst: GenericArray<u8, D>,
+) -> Result<GenericArray<u8, L>, InternalError>
+where
+    <D as Add<U1>>::Output: ArrayLength<u8>,
+{
+    let digest_len = <H as Digest>::OutputSize::USIZE;
+    let ell = div_ceil(L::USIZE, digest_len);
     if ell > 255 {
         return Err(InternalError::HashToCurveError);
     }
-    let dst_prime = [dst, &i2osp::<U1>(dst.len())?].concat();
+    let dst_prime = dst.concat(i2osp::<U1>(D::USIZE)?);
     let z_pad = i2osp::<<H as BlockInput>::BlockSize>(0)?;
-    let l_i_b_str = i2osp::<U2>(len_in_bytes)?;
-    let msg_prime = [
-        &z_pad,
-        msg,
-        &l_i_b_str,
-        i2osp::<U1>(0)?.as_slice(),
-        &dst_prime,
-    ]
-    .concat();
-
-    let mut b: Vec<Vec<u8>> = alloc::vec![H::digest(&msg_prime).to_vec()]; // b[0]
+    let l_i_b_str = i2osp::<U2>(L::USIZE)?;
+    let msg_0 = i2osp::<U1>(0)?;
+    let msg_prime =
+        core::array::IntoIter::new([z_pad.as_slice(), msg, &l_i_b_str, &msg_0, &dst_prime]);
 
     let mut h = H::new();
-    h.update(&b[0]);
-    h.update(&i2osp::<U1>(1)?);
-    h.update(&dst_prime);
-    b.push(h.finalize_reset().to_vec()); // b[1]
+    // b[0]
+    let b_0 = msg_prime
+        .into_iter()
+        .fold(&mut h, |h, msg| {
+            h.update(msg);
+            h
+        })
+        .finalize_reset();
+    let mut b_i = b_0.clone();
 
-    let mut uniform_bytes: Vec<u8> = Vec::new();
-    uniform_bytes.extend_from_slice(&b[1]);
+    let mut uniform_bytes = GenericArray::default();
 
-    for i in 2..(ell + 1) {
-        h.update(xor(&b[0], &b[i - 1])?);
-        h.update(&i2osp::<U1>(i)?);
+    for (i, chunk) in (1..(ell + 1)).zip(uniform_bytes.chunks_mut(digest_len)) {
+        if i == 1 {
+            h.update(b_0.clone());
+        } else {
+            h.update(xor(b_0.clone(), b_i.clone()));
+        }
```

Comment:
The cloning will be optimized away. I shorty contemplated to re-add the `Hash` trait to add an additional bind for `Copy`, I see at as a trade-off between cleaner code and user-facing API.

I don't like the `if` in the loop, but couldn't figure out a clean way to remove it. This loop is so small though, the compiler is sure to unroll it.

Reply:
Actually managed to remove the `if`:
https://github.com/novifinancial/voprf/blob/d7cb268c4e9ba7d24f8f02e04ff99e9e07dbd604/src/group/expand.rs#L54-L72